File: cloud/src/main/java/org/mozilla/msrp/platform/profile/FirefoxAccountService.java
package org.mozilla.msrp.platform.profile;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.json.JSONException;
import org.json.JSONObject;
import org.mozilla.msrp.platform.util.HttpUtil;
import org.springframework.stereotype.Service;

import javax.inject.Inject;
import java.io.IOException;

@Service
class FirefoxAccountService {

    @Inject
    FirefoxAccountServiceInfo firefoxAccountServiceInfo;

    String authorization(String code) throws IOException, JSONException {
        // token: fxCode -> fxToken
        JSONObject fxTokenJson = new JSONObject()
                .put("client_id", firefoxAccountServiceInfo.getClientId())
                .put("grant_type", "authorization_code")
                .put("ttl", 3600)
                .put("client_secret", firefoxAccountServiceInfo.getClientSecret())
                .put("code", code);

        String fxTokenRes = HttpUtil.post(firefoxAccountServiceInfo.getApiToken(), fxTokenJson, null);
        JSONObject fxJsonObject = new JSONObject(fxTokenRes);
        return fxJsonObject.getString("access_token");
    }

    FxAProfileResponse profile(String fxToken) throws JSONException, IOException {
        String fxaProfileRes = HttpUtil.post(firefoxAccountServiceInfo.getApiProfile(), null, fxToken);
        ObjectMapper mapper = new ObjectMapper();
        return mapper.readValue(fxaProfileRes, FxAProfileResponse.class);
    }

}

File: src/main/java/org/mozilla/msrp/platform/user/FirefoxAccountService.kt
package org.mozilla.msrp.platform.user

import org.json.JSONException
import org.springframework.stereotype.Service

import javax.inject.Inject
import java.io.IOException
import javax.inject.Named

/**
 * FirefoxAccountService prepare for the data to send to FirefoxAccountClient
 *
 * */
@Service
class FirefoxAccountService @Inject constructor(
    @Named("FxaAuth") var authClient: FirefoxAccountClient,
    @Named("FxaProfile") var profileClient: FirefoxAccountClient,
    private var firefoxAccountServiceInfo: FirefoxAccountServiceInfo) {


    @Throws(IOException::class, JSONException::class)
    fun token(fxaTokenRequest: FxaTokenRequest): String? {
        // token: fxCode -> fxToken

        val response = authClient.token(fxaTokenRequest).execute()
        val body = response.body()
        return if (response.isSuccessful && body != null) {
            body.access_token
        } else {
            null
        }
    }

    fun genFxaTokenRequest(authorizationCode: String): FxaTokenRequest {
        val clientId = firefoxAccountServiceInfo.clientId

        val clientSecret = firefoxAccountServiceInfo.clientSecret

        return FxaTokenRequest(
            clientId,
            clientSecret,
            authorizationCode)
    }

    @Throws(JSONException::class, IOException::class)
    fun profile(bearerHeader: String): FxaProfileResponse? {

        val response = profileClient.profile(bearerHeader).execute()

        val body = response.body()

        return if (response.isSuccessful) {
            body
        } else {
            null
        }

    }
}
